### PR TITLE
Cssproperties

### PR DESCRIPTION
**Pull request by bnwan**
This PR is based on the issue below:
#108

### CODE REVIEW

File: typings/css-properties.d.ts [hunk 1]
```diff
@@ -1525,5 +1525,5 @@ export interface CSSProperties {
   '::first-line'?: CSSProperties
   '::selection'?: CSSProperties
 
-  [propertyName: string]: string | number | CSSProperties | undefined
+  [propertyName: string]: string | number | boolean | Function | CSSProperties | undefined
```
Comment (ErikCupal) [inline typings/css-properties.d.ts]: Could you provide an example where this would be useful 🙂? I'm not sure whether it is really needed

**Comment by ErikCupal** [comment]
Thanks for PR 🙂. Yeah exporting the CSSProperties interface is not a bad idea, it might be useful for some usecases. It just isn't clear to me when would we need to assign `boolean` or `Function` to `[propertyName: string]`... Could you provide an example where this would be useful?

**Comment by bnwan** [comment]
Use case is when you want to pass a custom prop that is either a `boolean` or a `Function`. Obviously this is not a problem when using ES6, but with TypeScript they are not recognised. An example is below:

```
const Item = glamorous.li({
    borderWidth: 5,
    borderStyle: 'solid',
    marginRight: 5,
    marginBottom: 5,
    padding: 10,
    cursor: 'pointer'
  },
    (props: CSSProperties) => ({
      borderColor: props.isDefault? colors.actionColor : colors.borderColor,
    })
  );

 <Item
    key={doc.optionId}
    flagDefault={doc.flagDefault}
    onClick={() => onClick(doc)}
  >  
</Item>
```

**Comment by ErikCupal** [comment]
Aaah, I see.

There is no point in doing this. It can be done simply by generics. It should be in the docs, sorry for that. Here is an example with the generics:

```ts
const Item = glamorous.li<{ isDefault: boolean }>(
  {
    borderWidth: 5,
  },
  props => ({
    borderColor: props.isDefault ? 'red' : 'blue',
  })
)

const AnotherItem = () => (
  <Item isDefault={true}>
    Foo
  </Item >
)
```

This way you get 100% typechecking for custom dynamic properties. Look at this issue for more info #81
 🙂. 

Therefore the `boolean` and `Function` does not make any sense. Though I think that it can be useful to export the CSSProperties (e.g. when you need to create the object with styles alone and then pass its reference to glamorous).

**Comment by bnwan** [comment]
Ahhh...that makes sense...much better. Thank you 😄 

What about if you need to do something like this:

```
<glamorous.Div
   css={style}
   onClick={() => this.onSave()}
 >
 </glamorous.Div>
```

TypeScript complains about `onClick`. Is there a better way to do this?

**Comment by sammkj** [comment]
@bnwan the type definitions for built-in glamorous is incomplete, almost none I would say. To get rid of the `onClick` error, you need to add `React.HTMLAttributes<HTMLDivElement>` to `Div`.

**Comment by ErikCupal** [comment]
@bnwan I think we can close this PR. Is it ok 🙂?

**Comment by bnwan** [comment]
@ErikCupal Yes, but it'll be great if we can export `CSSProperties`

**Comment by kentcdodds** [comment]
> @ErikCupal Yes, but it'll be great if we can export CSSProperties

Let's make that happen in another PR :+1:

**Comment by bnwan** [comment]
@kentcdodds Great, I'll create another PR for it...thanks :)

**Comment by ErikCupal** [comment]
@bnwan Yup, I agree it would be good. Just for now: you can import the CSSProperties using `import { CSSProperties } from 'glamorous/typings/css-properties'` ;), though it is better when it is exported directly

**Pull request closed without merging**